File: erlang/xml.erl
-module(xml).
%%%
%%% A little library of routines to help with XML when talking to a GTH.
%%%

%% Generate tags
-export([attribute/2,
	 clip/1,
	 connection/4,
	 custom/2,
	 delete/1,
	 disable/1,
	 enable/2,
	 install/1,
	 job/1,
	 map/2,
	 new/3,
	 new_clip/1,
	 pcm_sink/2,
	 pcm_sink/3,
	 pcm_source/2,
	 player/3, player/4,
	 query_jobs/2, query_job/1,
	 query_resource/1,
	 recorder/4, recorder/5,
	 reset/1,
	 set/2,
	 tag/3, tag/2,
	 takeover/1,
	 tcp_sink/2, tcp_source/2,
	 unmap/1,
	 wide_recorder/4,
	 zero_job/1,
	 zero_resource/1]).

%%----------------------------------------------------------------------
%% XML generation

attribute(N, V) when is_list(N), is_list(V) ->
    tag("attribute", [{"name", N}, {"value", V}], "");

attribute(N, V) when is_list(N), is_integer(V) ->
    tag("attribute", [{"name", N}, {"value", integer_to_list(V)}], "").

clip(Id) when is_list(Id) ->
    tag("clip", [{"id", Id}], []).

connection(SP, ST, DP, DT) ->
    new("connection", [], [pcm_source(SP, ST), pcm_sink(DP, DT)]).

custom(Name, Attrs) when is_list(Name), is_list(Attrs) ->
    tag("custom", [{"name", Name}],
	[ attribute(N, stringify(V)) || {N, V} <- Attrs]).

disable(Name) ->
    tag("disable", [{"name", Name}]).

enable(Name, KVs) ->
    tag("enable", [{"name", Name}], [attribute(K, V) || {K, V} <- KVs]).

delete(Id) when is_list(Id) ->
    ["<delete id=\"", Id, "\"/>"].

install(Name) when is_list(Name) ->
    ["<install name=\"", Name, "\"/>"].

job(Id) when is_list(Id) ->
    ["<job id=\"", Id, "\"/>"].

map(_Target_type = pcm_source, Name) ->
    tag("map", [{"target_type", "pcm_source"}],
	tag("sdh_source", [{"name", Name}])).

new(Child, Attrs, Children) ->
    tag("new", [], tag(Child, Attrs, Children)).

new_clip(Id) when is_list(Id) ->
    tag("new", [], tag("clip", [{"id", Id}], [])).

player(Clip_list, Span, Timeslot) ->
    player(Clip_list, [], Span, Timeslot).

player(Clipnames, Attrs, Span, Timeslot)
  when is_list(Clipnames), is_list(hd(Clipnames)) ->
    Clips = [tag("clip", [{"id", Clipname}], "") || Clipname <- Clipnames],
    Sink = pcm_sink(Span, Timeslot),
    tag("new", [], tag("player", Attrs, [Clips, Sink])).

pcm_source(Span, Timeslot) when is_integer(Timeslot) ->
    ST = integer_to_list(Timeslot),
    tag("pcm_source", [{"span", Span}, {"timeslot", ST}], []).

pcm_sink(Span, Timeslot) when is_integer(Timeslot) ->
    ST = integer_to_list(Timeslot),
    tag("pcm_sink", [{"span", Span}, {"timeslot", ST}], []).

%% Experimental version for EBS. Experimental as at 2008-09-23
pcm_sink(IP, Span, Timeslot) when is_integer(Timeslot) ->
    ST = integer_to_list(Timeslot),
    tag("pcm_sink", [{"ip_addr", IP}, {"span", Span}, {"timeslot", ST}], []).

query_jobs(Ids, Verbose) ->
    Attrs = case Verbose of
		true -> [{"verbose", "true"}];
		false -> []
	    end,
    tag("query", Attrs, [tag("job", [{"id", Id}]) || Id <- Ids]).

query_job(Id) ->
    query_jobs([Id], false).

query_resource(Name) ->
    tag("query", [], tag("resource", [{"name", Name}])).

recorder(Span, Timeslot, Host, Port) ->
    recorder(Span, Timeslot, Host, Port, []).

recorder(Span, Timeslot, Host, Port, All_opts) ->
    Sink = case lists:member(udp, All_opts) of
	       true -> udp_sink(Host, Port);
	       false -> tcp_sink(Host, Port)
	   end,
    Opts = [X || X <- All_opts, X =/= udp],
    new("recorder", Opts, [pcm_source(Span, Timeslot), Sink]).

reset(Name) ->
    tag("reset", [], tag("resource", [{"name", Name}])).

set(Name, Attrs) when is_list(Name), is_list(Attrs) ->
    tag("set", [{"name", Name}],
	    [ attribute(N, stringify(V)) || {N, V} <- Attrs]).

stringify(S) when is_list(S) -> S;
stringify(N) when is_integer(N) -> integer_to_list(N);
stringify({A,B,C,D}) -> tl(lists:append(
			     [["."|integer_to_list(X)] || X <- [A,B,C,D]]));
stringify(B) when is_boolean(B) -> atom_to_list(B).

takeover(Jobs) ->
    tag("takeover", [], [job(Job) || Job <- Jobs]).

tcp_sink(IP, Port) when is_integer(Port) ->
    tag("tcp_sink", [{"ip_addr", IP}, {"ip_port", integer_to_list(Port)}]).

tcp_source(IP, Port) when is_integer(Port) ->
    tag("tcp_source", [{"ip_addr", IP}, {"ip_port", integer_to_list(Port)}]).

tag(Name, Attrs) ->
    tag(Name, Attrs, "").

tag(Name, Attrs, "") ->
    ["<", Name,
     [ [" ", N, "=\"", stringify(V), "\""] || {N, V} <- Attrs],
     "/>"];
tag(Name, Attrs, Child_text) ->
    ["<", Name,
     [ [" ", N, "=\"", stringify(V), "\""] || {N, V} <- Attrs],
     ">", Child_text, "</", Name, ">"].

unmap(Name) ->
    tag("unmap", [{"name", Name}]).

udp_sink(IP, Port) when is_integer(Port) ->
    tag("udp_sink", [{"ip_addr", IP}, {"ip_port", integer_to_list(Port)}]).

wide_recorder(Span, Host, Port, Tag) ->
    new("wide_recorder", [{"span", Span}, {"tag", Tag}],
	[udp_sink(Host, Port)]).


zero_job(Id) ->
    tag("zero", [], tag("job", [{"id", Id}])).

zero_resource(Name) ->
    tag("zero", [], tag("resource", [{"name", Name}])).

File: java/corelatus/gth/xml.java
//----------------------------------------------------------------------
// Title: XML tag generation
// Author: Matthias Lang (matthias@corelatus.com)
// Created: December 2020
//
// Copyright (c) 2020 Corelatus AB
//
// This is demonstration code. Use at your own risk.
//
// Generate XML for the Corelatus GTH API. We don't have code here for
// every part of the API---this is a useful subset for the sample programs.
//
//----------------------------------------------------------------------
package corelatus.gth;

//import java.util.AbstractMap;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

class xml {

    public static void main(String[] args)
    {
        System.out.println(xml.enable("pcm1A", make_map()));
    }

    public static String attrs_to_string(Map<String, String> attrs)
    {
        StringBuffer s = new StringBuffer("");
        attrs.forEach( (k, v) -> s.append(" " + k + "='" + v + "'") );
        return s.toString();
    }

    public static String attribute(String name, String value)
    {
        Map<String, String> map = make_map("name", name);
        map.put("value", value);
        return tag("attribute", map);
    }

    public static String clip(String id)
    {
        return tag("clip", make_map("id", id));
    }

    public static String custom(String name, Map<String, String> attrs)
    {
        return tag("custom", make_map("name", name), attrs_to_string(attrs));
    }

    public static String disable(String name)
    {
        return tag("disable", make_map("name", name));
    }

    public static String enable(String name, Map<String, String> attrs)
    {
        return tag("enable", make_map("name", name),
                   attribute_map_to_string(attrs));
    }

    public static String delete(String id)
    {
        return tag("delete", make_map("id", id));
    }

    public static String install(String name)
    {
        return tag("install", make_map("name", name));
    }

    public static String job(String id)
    {
        return tag("job", make_map("id", id));
    }

    public static String new_lapd_monitor(String ip, Integer port, String src)
    {
        return new_monitor("lapd_monitor", ip, port, src);
    }

    public static String map(String name)
    {
        return tag("map", make_map("target_type", "pcm_source"),
                   tag("sdh_source", make_map("name", name)));
    }

    public static String new_mtp2_monitor(String ip, Integer port, String src)
    {
        return new_monitor("mtp2_monitor", ip, port, src);
    }

    // <new> command is called 'g_new' here because 'new' is a keyword in Java
    public static String g_new(String children)
    {
        return tag("new", make_map(), children);
    }

    public static String pcm_source(String span, Integer timeslot)
    {
        Map<String, String> map = make_map("span", span);
        map.put("timeslot", timeslot.toString());
        return tag("pcm_source", map);
    }

    public static String pcm_source(String span,
                                    Integer timeslot,
                                    Integer first_bit,
                                    Integer bandwidth)
    {
        Map<String, String> map = make_map("span", span);
        map.put("timeslot", timeslot.toString());
        map.put("first_bit", first_bit.toString());
        map.put("bandwidth", bandwidth.toString());
        return tag("pcm_source", map);
    }

    public static String query_resource(String name)
    {
        return tag("query", make_map(), resource(name));
    }

    public static String query_job(String id)
    {
        return tag("query", make_map(),
                   tag("job", make_map("id", id)));
    }

    public static String recorder(String span,
                                  Integer timeslot,
                                  String host,
                                  Integer port)
    {
        return tag("new", make_map(),
                   tag("recorder", make_map(),
                       pcm_source(span, timeslot) + tcp_sink(host, port)));
    }

    public static String reset(String name)
    {
        return tag("reset", make_map(), resource(name));
    }

    public static String resource(String name)
    {
        return tag("resource", make_map("name", name));
    }

    public static String set(String name, Map<String, String> attrs)
    {
        return tag("set", make_map("name", name),
                   attribute_map_to_string(attrs));
    }

    public static String tag(String name, Map<String, String> attrs)
    {
        return "<" + name + attrs_to_string(attrs) + "/>";
    }

    public static String tag(String name,
                             Map<String, String> attrs,
                             String children)
    {
        return "<" + name + attrs_to_string(attrs) + ">" + children +
            "</" + name + ">";
    }

    public static String tcp_sink(String ip, Integer port)
    {
        Map<String, String> map = make_map("ip_addr", ip);
        map.put("ip_port", port.toString());
        return tag("tcp_sink", map);
    }

    public static String unmap(String name)
    {
        return tag("unmap", make_map("name", name));
    }

    public static String zero_job(String id)
    {
        return tag("zero", make_map(), tag("job", make_map("id", id)));
    }

    public static String zero_resource(String name)
    {
        return tag("zero", make_map(), resource(name));
    }

    //----------------------------------------------------------------------
    // Helpers

    public static Map<String, String> make_map()
    {
        return new HashMap<String, String>();
    }

    public static Map<String, String> make_map(String k, String v)
    {
        HashMap<String, String> map = new HashMap<String, String>();
        map.put(k, v);
        return map;
    }

    private static String attribute_map_to_string(Map<String, String> map)
    {
        StringBuffer sb = new StringBuffer();
        map.forEach( (k, v) -> sb.append(xml.attribute(k, v)) );
        return sb.toString();
    }

    private static String new_monitor(String type,
                                      String ip,
                                      Integer port,
                                      String source)
    {
        Map<String, String> map = make_map("ip_addr", ip);
        map.put("ip_port", port.toString());

        return g_new(tag(type, map, source));
    }

}
